### PR TITLE
refactor(key)!: use a builder-like pattern and owned internal representation

### PR DESCRIPTION
**Pull request by ssrlive**
I think the memory occupied by this structure cannot be very large. For ease of use, it is a good idea to remove its lifetime annotation.

### CODE REVIEW

**Comment by sseemayer** [comment]
Thanks for your contribution and welcome!

There is a currently ongoing discussion about redesigning the `DatabaseKey` in #143. Changing from borrowed to owned representation could be a part of that redesign, so I would prefer to combine your proposal with the outcome of the discussion so that backward compatibility is broken only once.

**Comment by ssrlive** [comment]
I think after making DatabaseKey clonable,  [adding support for other key elements such as OTP or hardware keys later](https://github.com/sseemayer/keepass-rs/issues/143#issuecomment-1427025702) will becoming more easer.

**Comment by sseemayer** [comment]
I've squashed together your commits and fixed up the commit message to pass our commit lint, then added a bit more refactoring on top to follow the suggestions from the #143 discussion, inspired by @simonsan 's idea of using a builder pattern (but not pulling database opening and saving logic into the builder and not fully following the pattern to a T). 

This new version should be cloneable and easily extensible with additional key element types in the future without more breaking changes in the future.

@varjolintu  and @louib, I would be happy to merge as-is, but would appreciate an additional review from at least one of you :)

**Comment by ssrlive** [comment]
How about change `withpassword` function like this:
```rust
    pub fn with_password(mut self, password: &str) -> Result<Self, DatabaseKeyError> {
        if password.is_empty() {
            return Err(DatabaseKeyError::IncorrectKey);
        }
        self.password = Some(password.to_string());
        Ok(self)
    }
```

File: README.md
Comment (sseemayer) [reply README.md]: Reusable `DatabaseKey`s are useful for settings where we want to load, process, and save a database. I like the idea of implementing `Zeroize` and have created #171 to put that onto the backlog.

Regarding a full database builder pattern, how would you imagine an API for saving databases to look like?

**Comment by sseemayer** [comment]
> For consistency, how about change `withpassword` function like this:
> 
> ```rust
>     pub fn with_password(mut self, password: &str) -> Result<Self, DatabaseKeyError> {
>         if password.is_empty() {
>             return Err(DatabaseKeyError::IncorrectKey);
>         }
>         self.password = Some(password.to_string());
>         Ok(self)
>     }
> ```
> 
> or like this:
> 
> ```rust
>     pub fn with_password(mut self, password: &str) -> Result<Self, DatabaseKeyError> {
>         if !password.is_empty() {
>             self.password = Some(password.to_string());
>         }
>         Ok(self)
>     }
> ```

I don't think returning an `Option` or `Result` type for `with_password` improves the code since passing an empty password should be valid, and `IncorrectKey` already has a different meaning. I have thus removed those commits from the PR again.

**Comment by ssrlive** [comment]
Hi @sseemayer , please merge this patch into the main branch yourself, so that it will be easier for me to submit new patches in the future.

**Comment by sseemayer** [comment]
> Hi @sseemayer , please merge this patch into the main branch yourself, so that it will be easier for me to submit new patches in the future.

I'll merge when it's ready. Particularly, I have still requested reviews from at least one of the other maintainers and we haven't made a decision if we want to combine this with the suggestion of further refactoring as suggested by @simonsan

**Comment by ssrlive** [comment]
I have not time to waiting for the approval, I will continue to adding contribute the project, welcoming to merge the branch https://github.com/ssrlive/keepass-rs/tree/refactor , this branch have merged all the changes recently.

**Comment by louib** [comment]
@sseemayer @ssrlive I wasn't sure if this PR was ready to review, which I why I didn't review it yet. From what I understand, there is still an outstanding issue. @sseemayer could this be addressed in a future PR? Otherwise please let me know when the PR is ready to review.

**Comment by ssrlive** [comment]
I decide to refactor the `Group / Entry` tree  using a `Node` trait. likes this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=df63308cd8f56f61e963649ae82198b5

I think it's better than `Node/Group/Entry/NodeRef/NodeRefMut`.

**Comment by ssrlive** [comment]
Hi @sseemayer  and @louib , I have rewritten the full project. using `Node trait / Group / Entry` architecture, please review it.
If you think it's necessary to merge it I will make a PR for it.
I think it will be easy to add features for the future. 
This is the link https://github.com/ssrlive/keepass-rs/tree/refactor

**Comment by varjolintu** [comment]
Sorry I haven't had time to review this yet. Going to do that soon.

**Comment by louib** [comment]
@ssrlive I believe I addressed some of the usability concerns in https://github.com/sseemayer/keepass-rs/pull/176
I had a look at the playground for the new interface and I'm not sure where the usability improvements are. Maybe I'm missing something. Either way, I think this is outside the scope of the current PR, which is focused on refactoring the `DatabaseKey` struct. Feel free to open an issue or another PR to discuss your improvement ideas.

**Review by louib — APPROVED** [review]

**Pull request merged**
